### DISCUSSION TITLE
Style Guide for Package code

### DISCUSSION THREAD

**Discussion by FullValueRider** [twinbasic/twinbasic#659] [Ideas]
We should no also open a discussion topic for guidance on how to write code for inclusion in a package so that we have consistency in how things are named and structured.

I remember that the Rubberduck project discussed such in one of its blogs for VBA.  Another example would be the Nim style guide, https://nim-lang.org/docs/nep1.html

My tuppeny starter would be that I use the following conventions in my code

Variables for Class state are corralled in a UDT called state and accessed using the variable s.
Variables for Class properties are corralled in a UDT called Properties and accessed using the variable p.

Variables declared inside a method are prefixed with 'my'

Parameters passed to a method are prefixed as follows
ip - input only
op - output only (value is returned through the parameter
iop - the passed value may be mutated and passed back via the parameter

The above may seem a bit pedantic but I've actually found the above 'rules' to be very useful in organising intellisense,  avoiding name clashes, and organising my thinking over what I'm doing with a particular variable.

Cheers

**Comment by wqweto** [comment]
Although this sounding very useful IMO these are not part of the VBx coding style/convention anyone would call idiomatic VBx code.

Hungarian notation although ingained in VBx code cannot be part of a modern coding style/convention too.

There are particular coding constructs which are desirable and others which should be avoided but I'm not sure firm rules can be established outside of a development team.

For instance I would like to see multiple statements on the same physical line (`a = 5: b = 10`) discouraged, single-line `If`s like in `If Something Then Other` discouraged, type characters like in `Dim a&` discouraged, variable declarations at the top of the routine preferred and so on.

**Comment by WaynePhillipsEA** [comment]
I think this would be quite difficult to get universal agreement on!

There's so much legacy VBx code out there that could now easily be put into our package manager and I worry that if we put up too many obstacles, people may be put off from contributing due to the effort required to refactor.

I'm interested in others opinions though.

**Reply by bclothier** [reply]
IMO, in long run, we will want one. 

However, we have to address the problem that there will be many legacy codebase that has its own particular style. I've already written elsewhere about how consistency is important because it removes the cognitive effort to parse someone else's code where the different style can distract or worse, hide the key element of how the code is supposed to run. 

We must also recognize that some will chafe at being required to conform to someone else's style and that can be just as big a barrier toward sharing code. They may even been developing using different tools. As an example, IDE did not help with the indentations so there were several indenter add-ins, each with their own style of indentations and in some cases, ability to customize how the code were to be indented. More variants, yay?

What we want to do is _evolve_ toward it. So my proposal is as follow:

1) Develop a style guide. 
2) Do not enforce it but have it be read & promoted.
3) Develop a linter.
4) For trivial changes, have the linter automatically make the changes.
5) For more complex changes, have the linter/analysis API flag the style issues.
6) At some point in future, make linter a prerequisite for submitting new package. 
7) Add more and more rule to the linter after a wide consensus from non-enforced style guide.

To use a more concrete example, let's consider code like:
```
Public Function fFizzBuzz(iNumber As Integer)
Dim x, s$

For x = 1 To iNumber
  If x Mod 15 = 0 Then
    s = s & vbCrLf & "FizzBuzz!"
  ElseIf x Mod 5 = 0 Then
   s = s & vbCrLf & "Buzz!"
  ElseIf x Mod 3 = 0 Then
    s = s & vbCrLf & "Fizz!"
  Else
    s = s & vbCrLf & x
  End If
Next

fFizzBuzz = Mid(s, 3)
End Function
```

In the first iteration of the style guide and linter, we can focus on trivial changes that makes it more clearer to read and make things more explicit:
```
Public Function fFizzBuzz(iNumber As Integer) As String
Dim x As Integer
Dim s As String

For x = 1 To iNumber
  If x Mod 15 = 0 Then
    s = s & vbCrLf & "FizzBuzz!"
  ElseIf x Mod 5 = 0 Then
   s = s & vbCrLf & "Buzz!"
  ElseIf x Mod 3 = 0 Then
    s = s & vbCrLf & "Fizz!"
  Else
    s = s & vbCrLf & x
  End If
Next

fFizzBuzz = Mid$(s, 3)
End Function
```

There are few more changes we can make without changing the code not covered by the FizzBuzz example:
```
cnn1 = cnn2
rs!Company
```
can be linted into:
```
cnn1.ConnectionString = cnn2.ConnectionString
rs.Fields("Company").Value
```
More verbose code but avoids misunderstanding on the reader's part which is more important at this point.

In the next iteration, the linter can then suggest to the author better names, and the author would need to agree. The renaming is done via refactoring so that it's very easy for the author to make changes without breaking or changing the code's behavior. The linter also can suggest a better constant to use, and author again decides:
```
Public Function FizzBuzz(Number As Integer) As String
    Dim x As Integer
    Dim s As String
    
    For x = 1 To Number
        If x Mod 15 = 0 Then
            s = s & vbNewLine & "FizzBuzz!"
        ElseIf x Mod 5 = 0 Then
            s = s & vbNewLine & "Buzz!"
        ElseIf x Mod 3 = 0 Then
            s = s & vbNewLine & "Fizz!"
        Else
            s = s & vbNewLine & x
        End If
    Next
    
    FizzBuzz = Mid$(s, 3)
End Function
```

In the next iteration, we can then suggest replacing magic numbers with constant for a more self-describing code. Again, the author has to agree and make changes. At this point, we are getting to the point where there will be more controversy because people can legitimately argue that this is less readable than the prior version, but there _also_ is a need to avoid magic numbers in a more complex codebase that the simple little FizzBuzz won't ever have.
```
Public Function FizzBuzz(Number As Integer) As String
    Dim x As Integer
    Dim s As String
    Const AfterLineBreak As Long = 3
    Const FizzBuzzNumber As Long = 15
    Const BuzzNumber As Long = 5
    Const FizzNumber As Long = 3
    Const FizzBuzzLine As String = "FizzBuzz!"
    Const BuzzLine As String = "Buzz!"
    Const FizzLine As String = "Fizz!"
    
    For x = 1 To Number
        If x Mod FizzBuzzNumber = 0 Then
            s = s & vbNewLine & FizzBuzzLine
        ElseIf x Mod BuzzNumber = 0 Then
            s = s & vbNewLine & BuzzLine
        ElseIf x Mod FizzNumber = 0 Then
            s = s & vbNewLine & FizzLine
        Else
            s = s & vbNewLine & x
        End If
    Next
    
    FizzBuzz = Mid$(s, AfterLineBreak)
End Function
```

It's my hope that by having a style guide that has a broad consensus, and aided by a linter & analysis API & refactoring tools, it will be then easy for the community to evolve and move toward a more uniform style guide which should take us somewhere between the 2nd and 3rd example.

**Comment by mwolfe02** [comment]
I'll just get my popcorn ready for the inevitable flame wars.  😁

**Comment by FullValueRider** [comment]
The style guide should focus things that cannot be adjusted on a per user basis.  Layout  is a per user issue as it comes from the prettyfier (and the options therein) and consequently would not be part of a style guide.

I'd agree with @wqweto that we don't want pointless hungarian notation.  The ones I suggested are useful in the VBA environment but are not required in the twinBasic environment.  This is because in the twinBasic environment hovering over a keyword pops up useful information.  In VBA it requires a right click and selection from a list to get the same information. My suggestions may still be useful because they are very simple, eliminate naming clashes and for parameters, more clearly state the programmer intended usage than byref and byval.
